core/src/main/scala/kafka/log/Log.scala

```diff
@@ -1572,12 +1572,28 @@ class Log(@volatile var dir: File,
         val offsetIdxFile = offsetIndexFile(dir, newOffset)
         val timeIdxFile = timeIndexFile(dir, newOffset)
         val txnIdxFile = transactionIndexFile(dir, newOffset)
-        for (file <- List(logFile, offsetIdxFile, timeIdxFile, txnIdxFile) if file.exists) {
-          warn(s"Newly rolled segment file ${file.getAbsolutePath} already exists; deleting it first")
-          Files.delete(file.toPath)
```

Comment:
This logic was very strange given the former validation below. I think I had not actually appreciated how it affected the analysis of this bug. If we had hit a case where we attempted to roll a segment that already existed in the `segments` map, then the log and index files would have been deleted by this logic and then recreated in `LogSegment.open`. Then we would have seen the failure in the call to `addSegment`. 

Is it possible that it was this sequence of events that caused us to observe an empty index file after the failure? In other words, could there be a cause other than an empty index that would trigger the unexpected roll?

Reply:
That’s true that the bug caused the log and index files to be deleted and then recreated. However, 2 out of about 4-5 times when this incident happened, I saw in the logs that the data log was loaded with 1 segment, log start offset X and log end offset X (log start == log end). Then, some time later, the first append happened to offset X which caused unexpected log roll and the exception.  So, we know that one of the causes is an empty index file. However, I don’t have proof that there is no other possible causes. (in other cases, I did not have full logs to be able to confirm empty index files)